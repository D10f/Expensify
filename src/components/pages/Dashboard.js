import React from 'react'
import ExpenseList from '../ExpenseList'
import ExpenseListFilters from '../ExpenseListFilters'

export const Dashboard = () => (
  <section>
    <header>These are your current expenses:</header>
    <ExpenseListFilters />
    <ExpenseList />
  </section>
)

export default Dashboard
