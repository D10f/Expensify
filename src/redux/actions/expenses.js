// import { v4 as uuid } from 'uuid'
import { database } from '../../firebase/firebase'

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

/*
 * Compared to 'regular' action generators that return an object
 * With redux thunk we can return also functions that get called immediately
 */

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    // Destructure from expenseData, with default values
    const {
      description = '',
      notes = '',
      amount = 100,
      createdAt = 0
    } = expenseData

    const expense = { description, amount, notes, createdAt }

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = (id) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}

export const editExpense = (id, updates = {}) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = []
        snapshot.forEach(expense => {
          expenses.push({
            id: expense.key,
            ...expense.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
  }
}
