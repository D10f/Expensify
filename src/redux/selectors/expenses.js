import moment from 'moment'

export default (expenses, { text, startDate, endDate, sortBy }) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? createdAtMoment.isSameOrAfter(startDate, 'day') : true
    const endDateMatch = endDate ? createdAtMoment.isSameOrBefore(endDate, 'day') : true
    // const startDateMatch = typeof startDate !== 'number' || expense.createdAt > startDate
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt < endDate
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    }

    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  })
}
