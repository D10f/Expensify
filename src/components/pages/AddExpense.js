import React from 'react'
import { connect } from 'react-redux'
import { addExpense } from '../../redux/actions/expenses'
import ExpenseForm from '../ExpenseForm'

export class AddExpensePage extends React.Component {

  onSubmit = (expense) => {
    this.props.addExpense(expense)
    this.props.history.push('/')
  }

  render(){
    return (
      <section>
        <h3>Create New Expense</h3>
        <ExpenseForm
          expense={{}}
          onSubmit={this.onSubmit} />
      </section>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispath(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
