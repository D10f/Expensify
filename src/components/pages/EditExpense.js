import React from 'react'
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../../redux/actions/expenses'
import ExpenseForm from '../ExpenseForm'

export class EditExpensePage extends React.Component {

  onClick = () => {
    this.props.removeExpense({ id: this.props.expense.id })
    this.props.history.push('/')
  }

  onSubmit = (updatedExpense) => {
    this.props.editExpense(this.props.expense.id, updatedExpense)
    this.props.history.push('/')
  }

  render(){
    return (
      <section>
        <button onClick={this.onClick}>Delete Expense</button>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
      </section>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, updates) => dispatch(editExpense(id, updates)),
  removeExpense: (id) => dispatch(removeExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
