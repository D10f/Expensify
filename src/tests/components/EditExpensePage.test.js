import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/pages/EditExpense'
import expenses from '../fixtures/expenses'

let wrapper, history, editExpense, removeExpense

beforeEach(() => {
  editExpense = jest.fn()
  removeExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<EditExpensePage history={history} expense={expenses[2]} editExpense={editExpense} removeExpense={removeExpense} />)
})

test('Should render correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Should handle onClick correctly', () => {
  wrapper.find('button').prop('onClick')()
  expect(removeExpense).toHaveBeenCalledWith({ id: expenses[2].id })
  expect(history.push).toHaveBeenCalledWith('/')
})

test('Should handle onSubmit correctly', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
  expect(editExpense).toHaveBeenCalledWith(expenses[2].id, expenses[2])
  expect(history.push).toHaveBeenCalledWith('/')
})
