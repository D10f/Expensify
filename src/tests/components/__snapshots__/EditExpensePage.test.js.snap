// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
<section>
  <button
    onClick={[Function]}
  >
    Delete Expense
  </button>
  <ExpenseForm
    expense={
      Object {
        "amount": 375,
        "createdAt": 1209600000,
        "description": "Chocolate",
        "id": "3",
        "note": "Dark chocolate",
      }
    }
    onSubmit={[Function]}
  />
</section>
`;
